pep-9999.rst

```diff
@@ -0,0 +1,520 @@
+  feature complete for the next X.Y.0 release
+* as with current alpha releases, all APIs added since the last CPython feature
+  release would be considered provisional
+* as with current alpha releases, beta releases would be prepared and published
+  from the master development branch
+
+
+Rather than being released monthly for a period of a few months while preparing
+a new X.Y.0 release, beta releases would instead be consistently published every
+other month.
+
+The only time this would not be the case is during the two month release
+candidate period for an upcoming X.Y.0 release.
+
+
+Changes to release candidate policy, phase duration, and cadence
+----------------------------------------------------------------
+
+Given the proposed changes to the beta release phase, the release candidate
+phase would see the following related adjustments:
+
+* Feature freeze, ABI freeze, pyc file format freeze, and maintenance branch
+  creation would all correspond with the creation of X.Y.0rc1 (currently these
+  occur across a mixture of X.Y.0b1, the last beta release, and X.Y.0rc1)
+* The X.Y.0 release candidate period would be extended from 3 weeks to 2 months
+* There would normally be two release candidates issued a month apart, but
+  additional candidates may be published at the release manager's discretion
+* The final release would occur between 1 and 4 weeks after the final release
+  candidate (depending if additional release candidates were needed after the
+  second)
+
+
+Removal of routine alpha releases
+---------------------------------
+
+Due to the proposed policy changes for beta releases and release candidates,
+CPython alpha releases would no longer serve a clear purpose, and as such
+would typically be skipped.
+
+If they chose to, release managers would still be free to make one or more
+alpha releases before entering the rolling beta release period. For example,
+that would be one way of testing out new release management tooling.
+
+
+Changes to extension module loading in CPython pre-releases
+-----------------------------------------------------------
+
+Extension modules built against CPython pre-releases prior to the ABI freeze
+for that release series may not actually load correctly on the final release.
+
+Historically, adoption of alpha and beta releases has been low enough that this
+hasn't really been a problem in practice. However, this PEP proposes to actively
+encourage widespread operational use of beta releases, and we'd like to make it
+clear that such binaries are only reliably portable to the exact version that
+was used to build them (they may fail to load on even the very next alpha or
+beta release).
+
+To that end, this PEP proposes amending the extension module ``SOABI`` marker
+for CPython pre-releases to use a "cpython_unstable-" prefix, and only switch
+back to the regular "cpython-" prefix once the ABI for that particular version
+has been frozen on entry to the release candidate stage.
```

Comment:
Why not a single-letter flag? Or a "b" suffix on the version (if our spec allows this...)

"Unstable" in particular carries different connotations for people, especially those who need security or legal audits of production code from people who may not be familiar with our use of the term.

Reply:
I've changed the working suggestion to be "cpython_prerelease" for now, but yes, that could be shortened even further and moved to appear as part of the version number.

`configure.ac` isn't set up to read from `patchlevel.h` though, and the ABI won't necessarily change on every new prerelease, so I think a single generic "this was built on a pre-release" qualifier should suffice (allowing rolling release consumers to make their own decisions on when they need to rebuild their extension modules)

Reply:
As I mentioned above (below? elsewhere!), the very next part of the tag will be `cpython-3.X` where X for prereleases already does not match the latest stable release. So the only advantage of changing this would be to prevent prerelease wheels from working unmodified with the same version when it becomes a stable release.

If that's the goal (which is valid, though I disagree with it), then we should mention it. But I think the version number being different already is going to be sufficient.

(And I haven't read throughout, but I assume stable ABI users get a different tag completely?)

Reply:
Yep, stable ABI users already get a different tag (abi3).

The issue being tackled here is only indirectly about wheel files though: it's addressing the fact that an extension module compiled on an alpha or beta release may segfault on the final release if it used a provisional or deprecated API that was removed in a later commit.

Historically we've been able to just ignore that, but if the betas are going to start being more widely used, we want to clearly mark extension modules that were built against a not-yet-stable version of the X.Y ABI.

Reply:
I've updated this section to cover Windows, and to use a short "p" pre-release ABI flag on both Windows and non-Windows systems.

I've also tried to clarify that the situation this is really aiming to avoid is users of the release candidates and stable releases getting bothered by segfaults arising from extension modules that were built against a not-yet-stable ABI.